File: mpd/VM/Platform.swift
// mpd — Mpd.VM.Platform namespace
// Reads/writes /var/lib/mpd/conf/platform.env — the in-VM identity file that
// records which kind of mpd setup this is:
//   MPD_PLATFORM=managed | sandbox
//   MPD_VM_IP=<ip>           (empty for sandbox; for managed it's the
//                              VM's static IP, e.g. 10.211.55.159)
//   MPD_VM_ID=<NNN>          3-digit VM identifier used everywhere:
//                              "100"-"254" for managed VMs (static-IP octet),
//                              "000" for the sandbox VM (DHCP — no fixed IP).
//                              Pod/container/hostname all build from this:
//                              `mpd-<NNN>` is the VM, `mpd-<NNN>-<runtime>`
//                              is each runtime inside it.
//
// Writers:
//   - managed VM via the host-side `mpd-virt` orchestrator (separate repo):
//     writes the file over SSH before `mpd --setup` runs in the VM.
//   - sandbox via setup/sandbox/lib/provision.sh: writes the file with
//     MPD_PLATFORM=sandbox / MPD_VM_ID=sandbox before `mpd --setup` runs.
//
// Reader: mpd's setup actions and helpers that need to know the platform
// or the VM ID at run-time. Lives under /var/lib/mpd/conf/ (persistent identity).

import Foundation

extension Mpd.VM.Platform {

    enum PlatformKind: String {
        case managed = "managed"
        case sandbox = "sandbox"
    }

    struct Identity {
        let platform: PlatformKind
        let vmIP: String
        let vmId: String   // 3-digit: "100"-"254" for managed, "000" for sandbox
    }

    /// Path to /var/lib/mpd/conf/platform.env.
    static var path: String {
        "\(Mpd.VM.confDir)/platform.env"
    }

    /// Load the identity file; throws with a fix-it message if missing.
    static func load() throws -> Identity {
        let fm = FileManager.default
        guard fm.fileExists(atPath: path) else {
            throw RuntimeError(
                "Missing \(path).\n" +
                "Run the matching bootstrap first:\n" +
                "  • sandbox VM:  setup/sandbox/take-over-sandbox-vm.sh\n" +
                "  • managed VM:  the host-side `mpd-virt` orchestrator")
        }

        let raw = (try? String(contentsOfFile: path, encoding: .utf8)) ?? ""
        let kv = parseKV(raw)

        guard let platformRaw = kv["MPD_PLATFORM"], let platform = PlatformKind(rawValue: platformRaw) else {
            throw RuntimeError("\(path): MPD_PLATFORM missing or invalid (expected: managed, sandbox).")
        }
        let vmIP = kv["MPD_VM_IP"] ?? ""
        let vmId = kv["MPD_VM_ID"] ?? ""

        return Identity(platform: platform, vmIP: vmIP, vmId: vmId)
    }

    /// Keys this writer manages. Other `MPD_*` keys (e.g. `MPD_NETWORK_*` set
    /// by a bootstrap script) are preserved verbatim by `write` so the
    /// bootstrap scripts and Platform can share the same file without
    /// clobbering each other.
    private static let managedKeys: Set<String> = [
        "MPD_PLATFORM", "MPD_VM_IP", "MPD_VM_ID",
    ]

    /// Write the identity file. Idempotent — overwrites the managed keys
    /// with the supplied values; preserves any other keys (e.g.
    /// `MPD_NETWORK_*`) that bootstrap scripts may have written.
    static func write(platform: PlatformKind, vmIP: String, vmId: String) throws {
        let fm = FileManager.default
        try fm.createDirectory(atPath: Mpd.VM.confDir, withIntermediateDirectories: true)

        // Collect non-managed keys from the existing file, in original order.
        var preserved: [(key: String, value: String)] = []
        if fm.fileExists(atPath: path) {
            let raw = (try? String(contentsOfFile: path, encoding: .utf8)) ?? ""
            for rawLine in raw.components(separatedBy: .newlines) {
                let line = rawLine.trimmingCharacters(in: .whitespaces)
                if line.isEmpty || line.hasPrefix("#") { continue }
                guard let eq = line.firstIndex(of: "=") else { continue }
                let key = String(line[..<eq])
                if managedKeys.contains(key) { continue }
                let value = String(line[line.index(after: eq)...])
                preserved.append((key, value))
            }
        }

        var body = """
            # mpd platform identity — written by setup, read at runtime.
            # Lives under /var/lib/mpd/conf/.
            MPD_PLATFORM=\(platform.rawValue)
            MPD_VM_IP=\(vmIP)
            # 3-digit VM identifier used in pod/container/hostname names.
            # Auto-derived from the VM hostname (mpd-<NNN>) at --setup; edit
            # to override. Runtime containers are named mpd-<NNN>-<runtime>.
            MPD_VM_ID=\(vmId)

            """
        if !preserved.isEmpty {
            body += "# Other keys preserved verbatim (set by bootstrap scripts):\n"
            for kv in preserved {
                body += "\(kv.key)=\(kv.value)\n"
            }
        }
        try body.write(toFile: path, atomically: true, encoding: .utf8)
    }

    /// Bootstrap helper: write the file if absent; leave any existing file
    /// untouched.
    static func ensureWritten(platform: PlatformKind, vmIP: String,
                              vmId: String) throws {
        if FileManager.default.fileExists(atPath: path) { return }
        try write(platform: platform, vmIP: vmIP, vmId: vmId)
    }

    /// Update only the VM ID in an existing platform.env. Used by
    /// `mpd --setup` to refresh the ID on every run (auto-derived from the
    /// VM hostname; the user can override by hand-editing the file).
    static func updateVmId(_ vmId: String) throws {
        let identity = try load()
        try write(platform: identity.platform, vmIP: identity.vmIP, vmId: vmId)
    }

    private static func parseKV(_ text: String) -> [String: String] {
        var result: [String: String] = [:]
        for rawLine in text.components(separatedBy: .newlines) {
            let line = rawLine.trimmingCharacters(in: .whitespaces)
            if line.isEmpty || line.hasPrefix("#") { continue }
            guard let eq = line.firstIndex(of: "=") else { continue }
            let key = String(line[..<eq])
            var value = String(line[line.index(after: eq)...])
            if value.hasPrefix("\""), value.hasSuffix("\""), value.count >= 2 {
                value = String(value.dropFirst().dropLast())
            }
            result[key] = value
        }
        return result
    }
}

File: go/internal/vm/platform.go
package vm

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

// PlatformEnvPath is the in-VM identity file. It records which kind of
// mpd install this is and which VM it is, and is read by everything that
// composes an address or a name.
const PlatformEnvPath = ConfDir + "/platform.env"

// PlatformKind distinguishes the two user-facing modes.
const (
	// PlatformManaged is a VM driven by the host-side mpd-virt
	// orchestrator; the user stays on their workstation.
	PlatformManaged = "managed"
	// PlatformSandbox is a desktop-inside-the-VM install; the user
	// lives in the VM.
	PlatformSandbox = "sandbox"
)

// PlatformIdentity is the content of platform.env that mpd owns.
type PlatformIdentity struct {
	// Platform is "managed" or "sandbox".
	Platform string
	// VMIP is the VM's static address, empty on sandbox (DHCP).
	VMIP string
	// VMID is the 3-digit identifier every name and address derives
	// from: "100"–"254" for managed VMs, "000" for the sandbox.
	VMID string
}

// managedKeys are the keys this writer owns. Everything else in the file
// — MPD_NETWORK_* written by a bootstrap script, say — is preserved
// verbatim, so the bootstrap scripts and mpd can share one file without
// clobbering each other.
var managedKeys = map[string]bool{
	"MPD_PLATFORM": true, "MPD_VM_IP": true, "MPD_VM_ID": true,
}

// LoadPlatform reads the VM's own platform.env.
func LoadPlatform() (PlatformIdentity, error) { return LoadPlatformFrom(PlatformEnvPath) }

// LoadPlatformFrom reads a platform.env at an explicit path, failing
// with the bootstrap command that creates it when it is absent.
func LoadPlatformFrom(path string) (PlatformIdentity, error) {
	body, err := os.ReadFile(path)
	if err != nil {
		return PlatformIdentity{}, fmt.Errorf("Missing %s.\n"+
			"Run the matching bootstrap first:\n"+
			"  • sandbox VM:  setup/sandbox/take-over-sandbox-vm.sh\n"+
			"  • managed VM:  the host-side `mpd-virt` orchestrator", path)
	}
	kv := parseEnv(string(body))
	platform := kv["MPD_PLATFORM"]
	if platform != PlatformManaged && platform != PlatformSandbox {
		return PlatformIdentity{}, fmt.Errorf(
			"%s: MPD_PLATFORM missing or invalid (expected: managed, sandbox).", path)
	}
	return PlatformIdentity{
		Platform: platform,
		VMIP:     kv["MPD_VM_IP"],
		VMID:     kv["MPD_VM_ID"],
	}, nil
}

// WritePlatform rewrites the VM's own platform.env.
func WritePlatform(id PlatformIdentity) error { return WritePlatformTo(PlatformEnvPath, id) }

// WritePlatformTo rewrites the managed keys at an explicit path and
// preserves every other line verbatim.
func WritePlatformTo(path string, id PlatformIdentity) error {
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}

	var preserved []string
	if body, err := os.ReadFile(path); err == nil {
		for _, raw := range strings.Split(string(body), "\n") {
			line := strings.TrimSpace(raw)
			if line == "" || strings.HasPrefix(line, "#") {
				continue
			}
			key, _, found := strings.Cut(line, "=")
			if !found || managedKeys[key] {
				continue
			}
			preserved = append(preserved, line)
		}
	}

	body := "# mpd platform identity — written by setup, read at runtime.\n" +
		"# Lives under /var/lib/mpd/conf/.\n" +
		"MPD_PLATFORM=" + id.Platform + "\n" +
		"MPD_VM_IP=" + id.VMIP + "\n" +
		"# 3-digit VM identifier used in pod/container/hostname names.\n" +
		"# Auto-derived from the VM hostname (mpd-<NNN>) at --vm-setup; edit\n" +
		"# to override. Runtime containers are named mpd-<NNN>-<runtime>.\n" +
		"MPD_VM_ID=" + id.VMID + "\n"
	if len(preserved) > 0 {
		body += "\n# Other keys preserved verbatim (set by bootstrap scripts):\n" +
			strings.Join(preserved, "\n") + "\n"
	}
	return os.WriteFile(path, []byte(body), 0o644)
}

// DeriveVMID reads the 3-digit VM identifier out of the VM's hostname.
//
// The hostname is the authority, not platform.env: it is what the
// hypervisor-side bootstrap set, it is what the user sees in their
// prompt, and a VM cloned to a new identity gets a new hostname. `mpd
// --vm-setup` re-derives on every run and writes the result back, so a
// hand-edited MPD_VM_ID survives only until the next setup.
//
// Returns "" when the hostname is not of the form mpd-<NNN>, which the
// caller reports rather than guessing at.
func DeriveVMID() string {
	// /etc/hostname directly rather than the hostname(1) binary: the
	// file is always present on Debian and needs no allow-list entry.
	body, err := os.ReadFile("/etc/hostname")
	if err != nil {
		return ""
	}
	return VMIDFromHostname(string(body))
}

// VMIDFromHostname extracts the identifier from a hostname string.
func VMIDFromHostname(raw string) string {
	host := strings.TrimSpace(raw)
	// Strip any FQDN form: only the short name carries the identifier.
	host, _, _ = strings.Cut(host, ".")
	id, found := strings.CutPrefix(host, "mpd-")
	if !found {
		return ""
	}
	return id
}

func parseEnv(text string) map[string]string {
	out := map[string]string{}
	for _, raw := range strings.Split(text, "\n") {
		line := strings.TrimSpace(raw)
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		key, value, found := strings.Cut(line, "=")
		if !found {
			continue
		}
		if len(value) >= 2 && strings.HasPrefix(value, `"`) && strings.HasSuffix(value, `"`) {
			value = value[1 : len(value)-1]
		}
		out[key] = value
	}
	return out
}
